File: src/Services/PollService/PollService.API/Models/Poll.cs
using System.ComponentModel.DataAnnotations.Schema;
using Common.Entities;

namespace PollService.API.Models;

public class Poll : BaseEntity<Guid>
{
    public required string Title { get; set; }
    public string? Topic { get; set; }
    public bool IsClosed { get; set; }
    public long DurationInSeconds { get; set; }

    public int UserId { get; set; }
    public UserCache User { get; set; } = null!;
    public ICollection<PollOption> PollOptions { get; set; } = [];

    [NotMapped]
    public TimeSpan Duration
    {
        get => TimeSpan.FromSeconds(DurationInSeconds);
        set => DurationInSeconds = (long)value.TotalSeconds;
    }

    [NotMapped]
    public DateTime EndTime => CreatedAt.UtcDateTime.Add(Duration);

    [NotMapped]
    public bool IsExpired => DateTime.UtcNow > EndTime;

    [NotMapped]
    public bool IsActive => !IsClosed && !IsExpired;
}

File: src/Client/VoteSystem.Client/src/app/_models/poll.ts
export interface Poll {
  id: string;
  title: string;
  topic: string;
  userId: number;
  userName: string;
  isClosed: boolean;
  duration: string;
  createdTime: Date;
  updatedTime: Date;
  endTime: Date;
  isExpired: boolean;
  isActive: boolean;
  pollOptions: PollOption[];
}
export interface PollOption {
  id: string;
  text: string;
  order: number;
  pollId: string;
  createdTime: Date;
  updatedTime?: Date;
}
export class CreatePollRequest {
  title: string;
  topic?: string;
  userId: number;
  timeCount: number;
  timeUnit: string;
  isClosed: boolean;
  pollOptionsToCreate: CreatePollOptionFromPollDto[];

  constructor(
    title: string,
    userId: number,
    isClosed: boolean,
    timeCount: number,
    timeUnit: string,
    pollOptionsToCreate: string[],
    topic?: string
  ) {
    const options = pollOptionsToCreate.map(
      (o) => new CreatePollOptionFromPollDto(o)
    );
    this.title = title;
    this.topic = topic;
    this.userId = userId;
    this.isClosed = isClosed;
    this.timeCount = timeCount;
    this.timeUnit = timeUnit;
    this.pollOptionsToCreate = options;
  }
}
class CreatePollOptionFromPollDto {
  text: string;
  constructor(text: string) {
    this.text = text;
  }
}
